### PR TITLE
 Make UploadedFile database id ('file' table) an integer sequence

### PR DESCRIPTION
**Pull request by sampierson**
This is a big changeset.  I would suggest not trying to review it all at once.  Review each commit individually.  Also note there is additional commentary in the commit messages to help you understand the commit.

This is not a perfect changeset.  It's always good to do the refactoring first then make your actual change.  Ideally I would reorder the commits to have the "Make UploadedFile database id ('file' table) an integer sequence" commit last, but doing that at this point would take about an extra day.

**REMINDER FOR SAM**: when you merge this you need to:
* Promote docker images to "latest"
* Update terraform tfvars: `csum_docker_image = "humancellatlas/upload-checksummer:7"`
* Get Ingest to rebuild their validators on `humancellatlas/upload-validator-base-alpine:18`

### CODE REVIEW

File: tests/functional/test_upload_service.py [hunk 1]
```diff
@@ -138,9 +138,12 @@ def _upload_area_record(self):
 
     def _checksum_record(self, filename):
         db = self.db_session_maker.session()
-        file_id = f"{self.upload_area_uuid}/{filename}"
-        record = db.query(DbChecksum).filter(DbChecksum.file_id == file_id).one_or_none()
-        return record
+        s3_key = f"{self.upload_area_uuid}/{filename}"
+        file_record = db.query(DbFile).filter(DbFile.s3_key == s3_key).one_or_none()
+        if file_record is None:
+            return None
```
Comment (parthshahva) [inline tests/functional/test_upload_service.py]: Should we raise an exception if we are trying to retrieve a checksum record for a file that doesn't exist

File: upload/common/uploaded_file.py [hunk 2]
```diff
@@ -29,39 +30,68 @@ def from_s3_key(cls, upload_area, s3_key):
...
+        self._s3obj = None
+        # utility:
+        self._db = UploadDB()
+
         if name and data and content_type:
-            self._create_s3_object(data, content_type)
-            self._fetch_or_create_db_record()
+            self._s3_create(name, data, content_type)
+            self._populate_properties_from_s3_object()
+            self._db_create()
```
Comment (parthshahva) [inline upload/common/uploaded_file.py]: do we need to call self._db_load() after this

File: upload/common/dss_checksums.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
+from functools import reduce
```
Comment (parthshahva) [inline upload/common/dss_checksums.py]: Nice!! This is a really clean interface

File: upload/common/uploaded_file.py [hunk 4]
```diff
@@ -17,16 +17,39 @@ class UploadedFile:
...
+        from .upload_area import UploadArea
+        upload_area = UploadArea(area_props['uuid'])
+        s3object = upload_area.s3_object_for_file(file_props['name'])
+        return cls(upload_area, s3object=s3object)
+
+    def __init__(self, upload_area, s3object):
+        """
+        The object of init() is to:
+        - populate properties from the S3 object
+        - create a DB record for this file of one does not exist
```
Comment (parthshahva) [inline upload/common/uploaded_file.py]: If __init__'s purpose is also this, why have a class create method?

File: upload/common/checksum_event.py [hunk 5]
```diff
@@ -9,13 +9,25 @@
...
+            file_id=prop_vals_dict['file_id'],
+            status=prop_vals_dict['status']
+        )
+
     def __init__(self, **kwargs):
         self.job_id = kwargs.get("job_id")
         self.id = kwargs["checksum_id"]
-        self.file_id = kwargs["file_id"]
-        self.status = kwargs["status"]
+        self.file_id = kwargs.get("file_id")
```
Comment (parthshahva) [inline upload/common/checksum_event.py]: Why this change? Would you ever initialize a checksum event without passing in the status and file id?

File: upload/common/validation_event.py [hunk 6]
```diff
@@ -10,10 +10,22 @@
...
+            file_id=prop_vals_dict['file_id'],
+            status=prop_vals_dict['status'],
+            docker_image=prop_vals_dict['docker_image']
+        )
+
     def __init__(self, **kwargs):
         self.id = kwargs["validation_id"]
         self.job_id = kwargs["job_id"]
-        self.file_id = kwargs["file_id"]
+        self.file_id = kwargs.get("file_id")
```
Comment (parthshahva) [inline upload/common/validation_event.py]: I'm probably missing something here. Why would we init a validation event without a file id?

Comment (parthshahva) [reply upload/common/validation_event.py]: Ah nevermind. This is from the docker image which doesn't actually update the event db record

File: tests/functional/test_upload_service.py (hunk 1 above)
Comment (sampierson) [reply tests/functional/test_upload_service.py]: It is sometimes the case in the functional test that the test gets ahead of the code being tested, and looks for records that haven't been committed yet.  The correct response in this case is to wait a second and try again.

File: upload/common/uploaded_file.py (hunk 2 above)
Comment (sampierson) [reply upload/common/uploaded_file.py]: `_db_create()` syncs the DB with the state of this object, *and* captures the id and inserts it back in `_properties`.  This is moot however as all this got refactored in a later commit.

File: upload/common/uploaded_file.py (hunk 4 above)
Comment (sampierson) [reply upload/common/uploaded_file.py]: The purpose of `create` is to create a new S3 object.
